### ISSUE TITLE
Parsing error for generic function with index signature return type

### ISSUE THREAD

**Issue by OliverJAsh** [eslint/typescript-eslint-parser#393]
**What version of TypeScript are you using?**
2.4.1

**What version of `typescript-eslint-parser` are you using?**
8

**What code were you trying to parse?**
``` ts
const x = <T>(): { [x: string]: string } => ({});
```

**What did you expect to happen?**
No error

**What happened?**
```
[eslint] Parsing error: ',' expected.
```

**Comment by JamesHenry** [comment]
I believe this behaviour is correct.

The source you have provided is genuinely a parse error in TypeScript 2.4.x:

http://astexplorer.net/#/gist/69f89d5c7455a7713684a53d49bdf7b1/67a4d6d0992298f9840e1b8fefe601c101ae3e40

(See the `parseDiagnostics` array)

In 2.5.x it seems to produce a valid AST with this parser. Please feel free to reopen if you have any more information.

**Comment by OliverJAsh** [comment]
The example is valid in TypeScript 2.4 and 2.5 only for `.ts` extensions—not when JSX is enabled, e.g. `.tsx`). I believe astexplorer.net is enabling JSX.

You can try it here: https://github.com/OliverJAsh/typescript-eslint-parser-parse-error-test

**Comment by OliverJAsh** [comment]
Here is an example of the error occurring inside a non-JSX (`.ts`) file: 
![image](https://user-images.githubusercontent.com/921609/31884065-dfc261f4-b7e3-11e7-8432-2953c62f7d1c.png)

**Comment by JamesHenry** [comment]
Ah yes you are right and this then is the same case as the one I dug into yesterday:

https://github.com/eslint/typescript-eslint-parser/issues/392#issuecomment-338495139

As outlined in that comment, you will need to disable `jsx` in your ESLint config or use constructs which are not ambiguous for TypeScript when in "JSX mode".

I have opened an issue to see if we can make the setting inferrable from file extension in the long run: https://github.com/eslint/typescript-eslint-parser/issues/399

**Comment by OliverJAsh** [comment]
Thank you, that makes sense!
